File: python/tests/test_sidecar.py
import re

import numpy as np
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
import pytest

from rts_gmlc_case import sidecar
from rts_gmlc_case.sidecar import SidecarWriter, data_hash

HASH_RE = re.compile(r"^[0-9a-f]{64}$")


def test_pinned_cross_language_hash():
    # Controller-verified literal (CAMPAIGN-FACTS.md canonical hash pin) —
    # the Julia sidecar implementation is pinned to the same value.
    assert data_hash(np.array([1.0, 2.0, 3.0])) == (
        "a68de4b5e96a60c8ceb3c7b7ef93461725bdbbff3516b136585a743b5c0ec664"
    )


def test_hash_is_64_lowercase_hex_chars():
    assert HASH_RE.match(data_hash(np.array([1.0, 2.0, 3.0])))


def test_integer_input_hashes_as_float64():
    # An all-integer array must be forced to float64 before hashing, or its
    # bytes differ from the float64 encoding of the same logical values.
    int_hash = data_hash(np.array([1, 2, 3]))
    float_hash = data_hash(np.array([1.0, 2.0, 3.0]))
    assert int_hash == float_hash


def _timestamps(n: int) -> pd.DatetimeIndex:
    return pd.date_range("2020-01-01", periods=n, freq="h", tz="UTC")


def test_dedup_same_values_twice_yields_one_file_same_uri(tmp_path):
    writer = SidecarWriter(tmp_path)
    values = np.array([1.0, 2.0, 3.0])
    timestamps = _timestamps(3)

    uri1, hash1 = writer.write(timestamps, values)
    uri2, hash2 = writer.write(timestamps, values)

    assert uri1 == uri2
    assert hash1 == hash2
    assert HASH_RE.match(hash1)

    files = list((tmp_path / "timeseries").glob("*.parquet"))
    assert len(files) == 1
    assert uri1 == f"timeseries/ts_{hash1[:16]}.parquet"


def test_distinct_values_yield_distinct_files(tmp_path):
    writer = SidecarWriter(tmp_path)
    timestamps = _timestamps(3)

    uri1, hash1 = writer.write(timestamps, np.array([1.0, 2.0, 3.0]))
    uri2, hash2 = writer.write(timestamps, np.array([4.0, 5.0, 6.0]))

    assert uri1 != uri2
    assert hash1 != hash2

    files = list((tmp_path / "timeseries").glob("*.parquet"))
    assert len(files) == 2


def test_round_trip_values_and_timestamp_type(tmp_path):
    writer = SidecarWriter(tmp_path)
    values = np.array([1.5, 2.5, 3.5])
    timestamps = _timestamps(3)

    uri, _ = writer.write(timestamps, values)

    table = pq.read_table(tmp_path / uri)
    assert table.schema.field("timestamp").type == pa.timestamp("ms", tz="UTC")
    assert table.schema.field("value").type == pa.float64()

    read_values = table.column("value").to_numpy()
    assert np.array_equal(read_values, values.astype("<f8"))

    read_ts = table.column("timestamp").to_pylist()
    expected_ts = list(timestamps.to_pydatetime())
    assert [t.replace(tzinfo=None) for t in read_ts] == [
        t.replace(tzinfo=None) for t in expected_ts
    ]


def test_write_returns_uri_relative_to_case_dir(tmp_path):
    writer = SidecarWriter(tmp_path)
    uri, data_hash_value = writer.write(_timestamps(3), np.array([9.0, 8.0, 7.0]))
    assert uri.startswith("timeseries/ts_")
    assert uri.endswith(".parquet")
    assert data_hash_value[:16] in uri


def test_corrupt_file_at_hash_path_is_replaced(tmp_path):
    # A prior interrupted run (or any other corruption) can leave garbage
    # bytes at the content-addressed path. `write` must not treat mere
    # existence as proof of a valid file — it must overwrite atomically so
    # the path is either absent or a complete, readable parquet file.
    writer = SidecarWriter(tmp_path)
    values = np.array([1.0, 2.0, 3.0])
    timestamps = _timestamps(3)

    digest = data_hash(values)
    corrupt_path = tmp_path / "timeseries" / f"ts_{digest[:16]}.parquet"
    corrupt_path.parent.mkdir(parents=True)
    corrupt_path.write_bytes(b"not a parquet file")

    uri, returned_hash = writer.write(timestamps, values)

    assert returned_hash == digest
    table = pq.read_table(tmp_path / uri)
    assert np.array_equal(table.column("value").to_numpy(), values.astype("<f8"))


def test_tz_naive_timestamps_raise(tmp_path):
    writer = SidecarWriter(tmp_path)
    naive = pd.date_range("2020-01-01", periods=3, freq="h")
    with pytest.raises(ValueError, match="tz"):
        writer.write(naive, np.array([1.0, 2.0, 3.0]))


def test_tz_aware_non_utc_converts_to_correct_instant(tmp_path):
    writer = SidecarWriter(tmp_path)
    eastern = pd.date_range("2020-01-01", periods=3, freq="h", tz="US/Eastern")
    values = np.array([1.0, 2.0, 3.0])

    uri, _ = writer.write(eastern, values)

    table = pq.read_table(tmp_path / uri)
    read_ts = table.column("timestamp").to_pylist()
    expected_utc = list(eastern.tz_convert("UTC").to_pydatetime())
    assert [t.replace(tzinfo=None) for t in read_ts] == [
        t.replace(tzinfo=None) for t in expected_utc
    ]


def test_mismatched_lengths_raise_with_both_lengths_named(tmp_path):
    writer = SidecarWriter(tmp_path)
    with pytest.raises(ValueError) as exc_info:
        writer.write(_timestamps(3), np.array([1.0, 2.0]))
    message = str(exc_info.value)
    assert "3" in message
    assert "2" in message
    assert not (tmp_path / "timeseries").exists()


def test_nan_payload_and_signed_zero_hash_as_raw_bytes():
    # Deliberate: this store hashes raw IEEE-754 bytes, not canonicalized
    # values. Two NaNs with different mantissa payloads, and +0.0/-0.0
    # (which compare equal), must hash differently.
    nan1 = np.array([np.nan], dtype="<f8")
    nan2 = nan1.copy()
    nan2.view(np.uint64)[0] ^= 1  # flip a low mantissa bit -> different NaN payload
    assert data_hash(nan1) != data_hash(nan2)

    assert data_hash(np.array([0.0])) != data_hash(np.array([-0.0]))


def test_write_failure_cleans_up_temp_file_and_propagates(tmp_path, monkeypatch):
    # Inject at os.replace, not pq.write_table: by the time os.replace runs,
    # the temp file has genuinely been written to disk, so this test can
    # actually observe whether the cleanup handler removes it. Injecting at
    # pq.write_table instead is vacuous -- the mock raises before any temp
    # file exists, so there is nothing to clean up and the test would pass
    # even with no cleanup code at all.
    def _boom(*args, **kwargs):
        raise RuntimeError("injected os.replace failure")

    monkeypatch.setattr(sidecar.os, "replace", _boom)

    writer = SidecarWriter(tmp_path)
    with pytest.raises(RuntimeError, match="injected os.replace failure"):
        writer.write(_timestamps(3), np.array([1.0, 2.0, 3.0]))

    leftovers = list((tmp_path / "timeseries").glob("*.tmp"))
    assert leftovers == []

File: julia/RTSGMLCCase/test/test_sidecar.jl
const _SIDECAR_HASH_RE = r"^[0-9a-f]{64}$"

function _sidecar_timestamps(n::Int)
    return [ZonedDateTime(Dates.DateTime(2020, 1, 1), tz"UTC") + Dates.Hour(i) for i in 0:(n - 1)]
end

@testset "pinned cross-language hash" begin
    # Controller-verified literal, confirmed directly against Python's
    # `rts_gmlc_case.sidecar.data_hash(np.array([1.0, 2.0, 3.0]))`.
    @test RTSGMLCCase.data_hash([1.0, 2.0, 3.0]) ==
          "a68de4b5e96a60c8ceb3c7b7ef93461725bdbbff3516b136585a743b5c0ec664"
end

@testset "hash is 64 lowercase hex chars" begin
    @test occursin(_SIDECAR_HASH_RE, RTSGMLCCase.data_hash([1.0, 2.0, 3.0]))
end

@testset "integer input hashes as float64" begin
    int_hash = RTSGMLCCase.data_hash(Float64.([1, 2, 3]))
    float_hash = RTSGMLCCase.data_hash([1.0, 2.0, 3.0])
    @test int_hash == float_hash
end

@testset "dedup same values twice yields one file same uri" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        values = [1.0, 2.0, 3.0]
        timestamps = _sidecar_timestamps(3)

        uri1, hash1 = RTSGMLCCase.write_series!(writer, timestamps, values)
        uri2, hash2 = RTSGMLCCase.write_series!(writer, timestamps, values)

        @test uri1 == uri2
        @test hash1 == hash2
        @test occursin(_SIDECAR_HASH_RE, hash1)

        files = filter(f -> endswith(f, ".parquet"), readdir(joinpath(dir, "timeseries")))
        @test length(files) == 1
        @test uri1 == "timeseries/ts_$(hash1[1:16]).parquet"
    end
end

@testset "distinct values yield distinct files" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        timestamps = _sidecar_timestamps(3)

        uri1, hash1 = RTSGMLCCase.write_series!(writer, timestamps, [1.0, 2.0, 3.0])
        uri2, hash2 = RTSGMLCCase.write_series!(writer, timestamps, [4.0, 5.0, 6.0])

        @test uri1 != uri2
        @test hash1 != hash2

        files = filter(f -> endswith(f, ".parquet"), readdir(joinpath(dir, "timeseries")))
        @test length(files) == 2
    end
end

@testset "round trip values and timestamp type" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        values = [1.5, 2.5, 3.5]
        timestamps = _sidecar_timestamps(3)

        uri, _ = RTSGMLCCase.write_series!(writer, timestamps, values)

        df = DataFrame(Parquet2.Dataset(joinpath(dir, uri)); copycols = true)
        @test eltype(df.timestamp) == Dates.DateTime
        @test eltype(df.value) == Float64
        @test df.value == values
        @test df.timestamp == Dates.DateTime.(timestamps)
    end
end

@testset "write returns uri relative to case dir" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        uri, hash_value = RTSGMLCCase.write_series!(writer, _sidecar_timestamps(3), [9.0, 8.0, 7.0])
        @test startswith(uri, "timeseries/ts_")
        @test endswith(uri, ".parquet")
        @test occursin(hash_value[1:16], uri)
    end
end

@testset "corrupt file at hash path is replaced" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        values = [1.0, 2.0, 3.0]
        timestamps = _sidecar_timestamps(3)

        digest = RTSGMLCCase.data_hash(values)
        corrupt_path = joinpath(dir, "timeseries", "ts_$(digest[1:16]).parquet")
        mkpath(dirname(corrupt_path))
        write(corrupt_path, "not a parquet file")

        uri, returned_hash = RTSGMLCCase.write_series!(writer, timestamps, values)

        @test returned_hash == digest
        df = DataFrame(Parquet2.Dataset(joinpath(dir, uri)); copycols = true)
        @test df.value == values
    end
end

@testset "non-UTC timestamps convert to the correct instant" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        eastern = [
            ZonedDateTime(Dates.DateTime(2020, 1, 1), tz"America/New_York") + Dates.Hour(i) for
            i in 0:2
        ]
        values = [1.0, 2.0, 3.0]

        uri, _ = RTSGMLCCase.write_series!(writer, eastern, values)

        df = DataFrame(Parquet2.Dataset(joinpath(dir, uri)); copycols = true)
        expected = Dates.DateTime.(astimezone.(eastern, tz"UTC"))
        @test df.timestamp == expected
    end
end

@testset "mismatched lengths raise with both lengths named" begin
    mktempdir() do dir
        writer = RTSGMLCCase.SidecarWriter(dir)
        err = try
            RTSGMLCCase.write_series!(writer, _sidecar_timestamps(3), [1.0, 2.0])
            nothing
        catch e
            e
        end
        @test err isa ErrorException
        @test occursin("3", err.msg)
        @test occursin("2", err.msg)
        @test !isdir(joinpath(dir, "timeseries"))
    end
end

@testset "nan payload and signed zero hash as raw bytes" begin
    nan1 = reinterpret(Float64, [reinterpret(UInt64, NaN)])
    nan2 = reinterpret(Float64, [reinterpret(UInt64, NaN) ⊻ UInt64(1)])
    @test RTSGMLCCase.data_hash(collect(nan1)) != RTSGMLCCase.data_hash(collect(nan2))
    @test RTSGMLCCase.data_hash([0.0]) != RTSGMLCCase.data_hash([-0.0])
end
